### PR TITLE
MHU: Add MHU driver support for SSE-200 subsystem

### PR DESCRIPTION
**Pull request by karl-zh**
MHU (Message Handling Unit) enable software to raise interrupts to
the processor cores.

This patch enable MHU driver on Musca A which has 2 MHU elements. It helps to communicate between 2 cores.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

### CODE REVIEW

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please look at implementing the IPM interface (`include/ipm.h`) rather than introducing a new interface.

**Comment by galak** [comment]
> Please look at implementing the IPM interface (`include/ipm.h`) rather than introducing a new interface.

Agreed, please use the IPM interface for this driver.

Also, do we need some code and support for bringing up the second core in Zephyr, and running Zephyr on it?

**Comment by karl-zh** [comment]
> > Please look at implementing the IPM interface (`include/ipm.h`) rather than introducing a new interface.
> 
> Agreed, please use the IPM interface for this driver.
> 
> Also, do we need some code and support for bringing up the second core in Zephyr, and running Zephyr on it?

Okay, updated to ipm interface. I was thought that the MHU has no data transfer compare to ipm.
Yes, it needs some change to wake up another core on musca, I will have a sample about use MHU on Musca.

File: drivers/ipm/ipm_mhu.c [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+#include <device.h>
+#include <soc.h>
+#include "ipm_mhu.h"
+
+#define DEV_CFG(dev) \
+	((const struct ipm_mhu_device_config * const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct ipm_mhu_data *)(dev)->driver_data)
+#define IPM_MHU_REGS(dev) \
+	((volatile struct ipm_mhu_reg_map_t  *)(DEV_CFG(dev))->base)
```
Comment (stanislav-poboril) [inline drivers/ipm/ipm_mhu.c]: two spaces between ipm_mhu_reg_map_t and *

File: drivers/ipm/ipm_mhu.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define IPM_MHU_MAX_DATA_SIZE		1
+#define IPM_MHU_MAX_ID_VAL		0
+#define SSE_200_CPU_ID_UNIT_OFFSET	((0x1F000UL))
+
+/* SSE 200 MHU register map structure */
+struct ipm_mhu_reg_map_t {
```
Comment (stanislav-poboril) [inline drivers/ipm/ipm_mhu.h]: Are these definitions somewhere in SoC's vendor SDK? If so, the vendor headers could be put into ext/hal/... and included and reused here (assuming there are no license issues).

**Review by stanislav-poboril — APPROVED** [review]
Looks good, I have some comments to the commit message:

Maybe it should start with "drivers/ipm: "?
Shouldn't the word "enables" be here: MHU (Message Handling Unit) enable software
And "It enabled in SSE 200 subsystems." seems to miss "is" perhaps.

File: drivers/ipm/Kconfig [hunk 3]
```diff
@@ -75,3 +75,21 @@ config IPM_IMX_MAX_ID_VAL
...
 	default 0 if IPM_IMX_MAX_DATA_SIZE_16
 	depends on IPM_IMX
+
+config IPM_MHU
+	bool "IPM MHU driver"
+	depends on IPM
+	help
+	  Driver for SSE 200 MHU (Message Handling Unit)
+
+config IPM_MHU0
```
Comment (galak) [inline drivers/ipm/Kconfig]: Do we need instance defines, can we get this from DT?

`DT_ARM_MHU_0` or `DT_ARM_MHU_1`

File: drivers/ipm/ipm_mhu.c [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+#if defined(CONFIG_IPM_MHU0) || defined(CONFIG_IPM_MHU1)
```
Comment (galak) [inline drivers/ipm/ipm_mhu.c]: This ifdef shouldn't be needed.

File: drivers/ipm/ipm_mhu.c [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+static enum ipm_mhu_cpu_id_t ipm_mhu_get_cpu_id(const struct device *d)
+{
+	volatile u32_t *p_mhu_dev_base;
+	volatile u32_t *p_cpu_id;
+
+	p_mhu_dev_base = (volatile u32_t *)IPM_MHU_REGS(d);
+
+	p_cpu_id = (volatile u32_t *)(((u32_t)p_mhu_dev_base & 0xF0000000) +
```
Comment (galak) [inline drivers/ipm/ipm_mhu.c]: Can we add a #define for the magic number 0xF0000000

File: drivers/ipm/ipm_mhu.h (hunk 2 above)
Comment (galak) [reply drivers/ipm/ipm_mhu.h]: The ARM IP has followed this model of not having a SDK or headers, so its fine that we define the register map here.

File: boards/arm/v2m_musca/v2m_musca-common.dtsi [hunk 6]
```diff
@@ -22,6 +22,18 @@ dtimer0: dtimer@2000 {
 	interrupts = <5 3>;
 };
 
+mhu0: mhu@3000 {
```
Comment (galak) [inline boards/arm/v2m_musca/v2m_musca-common.dtsi]: We should add `label` properties, its the current way we use in Zephyr to kinda map device_get_binding() and DTS.

File: dts/bindings/mhu/arm,mhu.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    This binding gives a base representation of the ARM MHU
+
+properties:
+    compatible:
+      type: string
+      category: required
+      description: compatible strings
+      constraint: "arm,mhu"
+      generation: define
+
```
Comment (galak) [inline dts/bindings/mhu/arm,mhu.yaml]: add a label property here

**Review by galak — CHANGES_REQUESTED** [review]
Looks good, some minor review comments in a few places.

File: drivers/ipm/ipm_mhu.h (hunk 2 above)
Comment (stanislav-poboril) [reply drivers/ipm/ipm_mhu.h]: Ok. Btw I think checkpatch.pl ignores the ext folder?

File: drivers/ipm/Kconfig (hunk 3 above)
Comment (karl-zh) [reply drivers/ipm/Kconfig]: DT_ARM_MHU_0 and DT_ARM_MHU_1 are always defined after mhu added into dts file, are they?
The config here provide a choice which MHU device should add as in this sub system include 2 MHU.
When enabling MHU device, there will be settings in the prj.conf like:

CONFIG_IPM=y
CONFIG_IPM_MHU=y
CONFIG_IPM_MHU0=y
CONFIG_IPM_MHU1=y

File: drivers/ipm/ipm_mhu.c (hunk 4 above)
Comment (karl-zh) [reply drivers/ipm/ipm_mhu.c]: CONFIG_IPM=y
CONFIG_IPM_MHU=y
#CONFIG_IPM_MHU0=y
#CONFIG_IPM_MHU1=y

If config file above, there are warnings about some symbols declared but not used. Do we need to consider this scenario?

File: dts/bindings/mhu/arm,mhu.yaml (hunk 7 above)
Comment (karl-zh) [reply dts/bindings/mhu/arm,mhu.yaml]: Added the MHU label, and use it in the code, commit 51ac287cffd90214aef115b649a71fda6873f669.
Thanks

File: samples/subsys/ipc/ipm_mhu_dual_core/README.rst [hunk 8]
```diff
@@ -0,0 +1,48 @@
+.. _ipm_mhu_dual_core:
+
+MHU Dual Core
+###########
```
Comment (dbkinder) [inline samples/subsys/ipc/ipm_mhu_dual_core/README.rst]: The heading underline must be at least as long as the title, so add a couple more ``##``

File: samples/subsys/ipc/ipm_mhu_dual_core/README.rst [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+Overview
+********
+A simple dual core example that can be used with arm Musca A1
```
Comment (dbkinder) [inline samples/subsys/ipc/ipm_mhu_dual_core/README.rst]: After reading this description I have no idea what this sample does or what it's for (but you do mention some facts in the commit message that should be added here.

How about:
```
An MHU (Message Handling Unit) enables software to raise interrupts to
the processor cores.  This sample is a simple dual-core example for a
Musca A1 board that has two MHU units.
```

And then can you add some more explanation of what the sample does?

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: samples/subsys/ipc/ipm_mhu_dual_core/README.rst (hunk 9 above)
Comment (karl-zh) [reply samples/subsys/ipc/ipm_mhu_dual_core/README.rst]: Thanks, I have updated the documents in latest commit.

File: samples/subsys/ipc/ipm_mhu_dual_core/README.rst [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+MHU Dual Core
+#############
+
+Overview
+********
+An MHU (Message Handling Unit) enables software to raise interrupts to
+the processor cores. This sample is a simple dual-core example for a
+Musca A1 board that has two MHU units. This sample only test MHU0, the
+steps are:
+1. CPU 0 will wake up CPU 1 after initialization
```
Comment (dbkinder) [inline samples/subsys/ipc/ipm_mhu_dual_core/README.rst]: add a blank line before the first item in a list

File: samples/subsys/ipc/ipm_mhu_dual_core/README.rst [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+2. CPU 1 will send to CPU 0 an interrupt over MHU0
+3. CPU 0 return the same to CPU 1 when received MHU0 interrupt
+4. Test done when CPU 1 received MHU0 interrupt
+
+* single thread
```
Comment (dbkinder) [inline samples/subsys/ipc/ipm_mhu_dual_core/README.rst]: Not sure what this bullet point is for?  Delete it?

**Review by dbkinder — CHANGES_REQUESTED** [review]
getting closer on the docs...

**Comment by karl-zh** [comment]
> getting closer on the docs...

Thanks, updated now.

**Review by dbkinder — APPROVED** [review]

File: drivers/ipm/ipm_mhu.c [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	return IPM_MHU_ERR_NONE;
+}
+
+static int ipm_mhu_send(struct device *d, int wait, u32_t cpu_id,
+			  const void *data, int size)
+{
+	ARG_UNUSED(wait);
+	u32_t set_val = *(u32_t *)data;
```
Comment (karl-zh) [inline drivers/ipm/ipm_mhu.c]: The register value should not from the param.
Suggest to use a fixed number instead:

	ARG_UNUSED(data);
	u32_t set_val = 0x01;

File: drivers/ipm/Kconfig (hunk 3 above)
Comment (galak) [reply drivers/ipm/Kconfig]: But that choice of which MHU can and should be handled by whats in the DTS.   I don't think we need the Kconfig options here.

File: samples/subsys/ipc/ipm_mhu_dual_core/sample.yaml [hunk 13]
```diff
@@ -0,0 +1,8 @@
+sample:
+  description: MHU dual core sample
+    application
+  name: ipm_mhu_dual_core
+tests:
+  test:
+    tags: ipm
+    platform_whitelist: arm Musca A1 board
```
Comment (galak) [inline samples/subsys/ipc/ipm_mhu_dual_core/sample.yaml]: This doesn't look right.

File: samples/subsys/ipc/ipm_mhu_dual_core/src/main.c [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+#define SSE_200_SYSTEM_CTRL_CPU_WAIT	(SSE_200_SYSTEM_CTRL_S_BASE + 0x118)
+#define SSE_200_CPU_ID_UNIT_BASE	(0x5001F000UL)
+
+#define NON_SECURE_FLASH_ADDRESS	(192 * 1024)
+#define NON_SECURE_IMAGE_HEADER		(0x400)
+#define NON_SECURE_FLASH_OFFSET		(0x10000000)
+
+struct device *mhu0;
+
+static void wakeup_cpu1(void)
```
Comment (MaureenHelm) [inline samples/subsys/ipc/ipm_mhu_dual_core/src/main.c]: Why isn't this in the soc or board-level support?

Comment (karl-zh) [reply samples/subsys/ipc/ipm_mhu_dual_core/src/main.c]: @galak Hi Kumar,
What is your opinion on that? The driver would be in ext/hal/xxx folder for each platform, right? Current Zephyr seems no arm folder in ext/hal/.

File: drivers/ipm/ipm_mhu.c [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+	.send = ipm_mhu_send,
+	.register_callback = ipm_mhu_register_cb,
+	.max_data_size_get = ipm_mhu_max_data_size_get,
+	.max_id_val_get = ipm_mhu_max_id_val_get,
+	.set_enabled = ipm_mhu_set_enabled,
+};
+
+static void ipm_mhu_irq_config_func_0(struct device *d);
+
+static struct ipm_mhu_device_config ipm_mhu_cfg_0 = {
```
Comment (MaureenHelm) [inline drivers/ipm/ipm_mhu.c]: This can be `const`

File: drivers/ipm/ipm_mhu.c [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+			DT_ARM_MHU_0_IRQ_0,
+			ipm_mhu_isr,
+			DEVICE_GET(mhu_0),
+			0);
+	irq_enable(DT_ARM_MHU_0_IRQ_0);
+}
+
+static void ipm_mhu_irq_config_func_1(struct device *d);
+
+static struct ipm_mhu_device_config ipm_mhu_cfg_1 = {
```
Comment (MaureenHelm) [inline drivers/ipm/ipm_mhu.c]: This can also be `const`

File: dts/bindings/mhu/arm,mhu.yaml [hunk 17]
```diff
@@ -0,0 +1,33 @@
+---
```
Comment (MaureenHelm) [inline dts/bindings/mhu/arm,mhu.yaml]: Add a license/copyright header

File: samples/subsys/ipc/ipm_mhu_dual_core/CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.13.1)
```
Comment (MaureenHelm) [inline samples/subsys/ipc/ipm_mhu_dual_core/CMakeLists.txt]: Add a license/copyright header

File: samples/subsys/ipc/ipm_mhu_dual_core/src/main.c (hunk 14 above)
Comment (MaureenHelm) [reply samples/subsys/ipc/ipm_mhu_dual_core/src/main.c]: Why would it be in `ext/`? I think it belongs in `soc/arm/arm/musca_a`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/ipm/ipm_mhu.c [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+	.send = ipm_mhu_send,
+	.register_callback = ipm_mhu_register_cb,
+	.max_data_size_get = ipm_mhu_max_data_size_get,
+	.max_id_val_get = ipm_mhu_max_id_val_get,
+	.set_enabled = ipm_mhu_set_enabled,
+};
+
+static void ipm_mhu_irq_config_func_0(struct device *d);
+
+const struct ipm_mhu_device_config ipm_mhu_cfg_0 = {
```
Comment (MaureenHelm) [inline drivers/ipm/ipm_mhu.c]: This should be `static` AND `const`

File: drivers/ipm/ipm_mhu.c [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+			DT_ARM_MHU_0_IRQ_0,
+			ipm_mhu_isr,
+			DEVICE_GET(mhu_0),
+			0);
+	irq_enable(DT_ARM_MHU_0_IRQ_0);
+}
+
+static void ipm_mhu_irq_config_func_1(struct device *d);
+
+const struct ipm_mhu_device_config ipm_mhu_cfg_1 = {
```
Comment (MaureenHelm) [inline drivers/ipm/ipm_mhu.c]: `static const`

File: samples/subsys/ipc/ipm_mhu_dual_core/src/main.c [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+	MHU_CPU0 = 0,
+	MHU_CPU1,
+	MHU_CPU_MAX,
+};
+
+struct device *mhu0;
+
+extern void wakeup_cpu1(void);
+
+extern u32_t sse_200_platform_get_cpu_id(void);
```
Comment (MaureenHelm) [inline samples/subsys/ipc/ipm_mhu_dual_core/src/main.c]: Move these externs to `soc.h`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: samples/subsys/ipc/ipm_mhu_dual_core/src/main.c (hunk 21 above)
Comment (karl-zh) [reply samples/subsys/ipc/ipm_mhu_dual_core/src/main.c]: Fixed.

**Review by MaureenHelm — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Pull request merged**